### PR TITLE
PCHR-2456: Changes to table of documents on documents dashboard

### PR DESCRIPTION
**Pull request by bmnepali**
## Overview
This PR adds new columns in the document list and makes them sortable when clicked on them.

## Before
- List document without 'assignees' column, target contact column name is not correct and the columns are not sortable on click.
screen shot 2017-09-11 at 7 44 41 pm

## After
- Adds 'assignees" column, target contact column name changes to staff and make the column sortable.
sorting-min

## Technical Details
1. Created  a function to sort the document list bases on the property name to sort by.
```js
$scope.propertyName = 'activity-date-time';
$scope.reverse = true;

/**
 * Sort the document list based on the property type
 * @param  {string} propertyName
 * @return {array}
 */
$scope.sortBy = function (propertyName) {
  $scope.reverse = ($scope.propertyName === propertyName) ? !$scope.reverse : false;
  $scope.propertyName = propertyName;

  switch (propertyName) {
    case 'type':
      $scope.list = _.sortBy($scope.list, function (doc) {
        return $rootScope.cache.documentType.obj[doc.activity_type_id];
      });
      break;
    case 'status_id':
      $scope.list = _.sortBy($scope.list, function (doc) {
        return $rootScope.cache.documentStatus.obj[doc.status_id];
      });
      break;
    case 'target_contact':
      $scope.list = _.sortBy($scope.list, function (doc) {
        return $rootScope.cache.contact.obj[doc.target_contact_id[0]].sort_name;
      });
      break;
    case 'assignee':
      $scope.list = _.sortBy($scope.list, function (doc) {
        var assignee = doc.assignee_contact_id.length && _.find($rootScope.cache.contact.obj, {'id': doc.assignee_contact_id[0]});

        return assignee && assignee.sort_name;
      });
      break;
    case 'case_id':
      $scope.list = _.sortBy($scope.list, function (doc) {
        var assignment = $rootScope.cache.assignment.obj[doc.case_id];
        var assignmentType = assignment && $rootScope.cache.assignmentType.obj[assignment.case_type_id];

        return assignmentType && assignmentType.title;
      });
      break;
  }
};
```
2.  Added a function to create a list of assignees form their ids.
```js
/**
 * Creates the list of contact name as  object
 * @param  {array} assigneesIds
 * @return {object}
 */
$scope.listAssignees = function (assigneesIds) {
  var assigneeList = {};

  if (assigneesIds.length) {
    _.each(assigneesIds, function (assigneeId) {
      var assignee = _.find($rootScope.cache.contact.obj, {'contact_id': assigneeId});

      if (assignee) {
        assigneeList[assigneeId] = assignee.sort_name.replace(',', '');
      }
    });
  }

  return assigneeList;
};
```
3.  Add styles for dropdown arrow in column headers uk.co.compucorp.civicrm.tasksassignments/scss/civitasks/partials/_documents.scss
```css
table th {
  cursor: pointer;
}

.sortorder:after {
  content: '\25b2';
}
.sortorder.reverse:after {
  content: '\25bc';
}
```
4.  Change document list table header to add column header sort feature on click
```html
<thead>
  <tr>
    <th ng-click="sortBy('type')">
      <span class="sortorder" ng-if="propertyName === 'type'" ng-class="{reverse: reverse}"></span>
      Type
    </th>
    <th ng-click="sortBy('target_contact_id')">
      <span class="sortorder" ng-if="propertyName === 'target_contact_id'" ng-class="{reverse: reverse}"></span>
      Staff
    </th>
    <th ng-click="sortBy('assignee')">
      <span class="sortorder" ng-if="propertyName === 'assignee'" ng-class="{reverse: reverse}"></span>
      Assignee
    </th>
    <th ng-click="sortBy('case_id')" ng-if="settings.extEnabled.assignments">
      <span class="sortorder" ng-if="propertyName === 'case_id'" ng-class="{reverse: reverse}"></span>
      Assignment
    </th>
    <th ng-click="sortBy('activity-date-time')">
      <span class="sortorder" ng-if="propertyName === 'activity-date-time'" ng-class="{reverse: reverse}"></span>
      Due Date
    </th>
    <th ng-click="sortBy('expire_date')">
      <span class="sortorder" ng-if="propertyName === 'expire_date'" ng-class="{reverse: reverse}"></span>
      Expiry Date
    </th>
    <th ng-click="sortBy('status_id')">
      <span class="sortorder" ng-show="propertyName === 'status_id'" ng-class="{reverse: reverse}"></span>
      Status
    </th>
    <th></th>
  </tr>
</thead>
```
5.  Make list of assignees clickable.
```html
<td ng-if="document.assignee_contact_id">
    <a ng-repeat="(key, value) in listAssignees(document.assignee_contact_id)" ng-href="{{url.CONTACT+'?reset=1&cid='+cache.contact.obj[key].contact_id}}">
      {{value}} {{$last ? '' : ', '}}
    </a>
  </td>
```
## Comments

- As part of this Pr also, some mock data for documents, assignments and contacts have been updated and added as per the requirements of this ticket.

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js [hunk 1]
```diff
@@ -21,6 +22,10 @@ define([
 
       $scope.dueThisWeek = 0;
       $scope.dueToday = 0;
+
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js]: Why the empty lines. Please remove them.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js [hunk 2]
```diff
@@ -74,6 +79,70 @@ define([
       };
 
       /**
+       * Sort the document list based on the property type
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js]: Should not it be `property name`?

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js [hunk 3]
```diff
@@ -312,7 +312,7 @@ define([
           contactIds.push(document.source_contact_id);
 
           if (document.assignee_contact_id && document.assignee_contact_id.length) {
-            contactIds.push(document.assignee_contact_id[0]);
+            Array.prototype.push.apply(contactIds, document.assignee_contact_id);
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: Cant we can simply push using `contactIds.push(document.assignee_contact_id)`?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 4]
```diff
@@ -2,15 +2,17 @@
 /* eslint-env amd, jasmine */
 
 define([
-  'common/moment',
   'common/angular',
+  'common/moment',
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Alphabetical order is not maintained.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 5]
```diff
@@ -2,15 +2,17 @@
...
+  'common/lodash',
   'mocks/fabricators/document',
+  'mocks/fabricators/contact',
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Alphabetical order is not maintained.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 6]
```diff
@@ -143,8 +145,136 @@ define([
...
+          $rootScope.cache.contact.obj[option.contact_id] = option;
+        });
+
+        $rootScope.cache.assignmentType.obj = assignmentFabricator.assignmentTypes();
+        $rootScope.cache.assignment.obj = assignmentFabricator.listAssignments();
+      });
+
+      describe('documents are not sorted', function () {
+        it('lists unsorted document list', function () {
+          expect($scope.list[0].id).toBe('1200');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: This test is not clear to me, why are we testing with two arbitrary number 1200 and 1213.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 7]
```diff
@@ -143,8 +145,136 @@ define([
...
+        beforeEach(function () {
+          $scope.sortBy('type');
+        });
+
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+
+        it('lists documents by types', function () {
+          expect($scope.list[0].id).toBe('1213');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Same as above.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 8]
```diff
@@ -143,8 +145,136 @@ define([
...
+      describe('documents are sorted by document status', function () {
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+        beforeEach(function () {
+          $scope.sortBy('status_id');
+        });
+
+        it('lists documents by document status', function () {
+          expect($scope.list[0].id).toBe('1210');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Same as above.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 9]
```diff
@@ -143,8 +145,136 @@ define([
...
+          expect($scope.list[0].id).toBe('1213');
+          expect($scope.list[4].id).toBe('1200');
+        });
+      });
+
+      describe('documents are sorted by document status', function () {
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+        beforeEach(function () {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Give  a line break.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 10]
```diff
@@ -143,8 +145,136 @@ define([
...
+        afterEach(function () {
+          $rootScope.$apply();
         });
+
+        beforeEach(function () {
+          $scope.sortBy('target_contact');
+        });
+
+        it('lists documents by target contact/staff ', function () {
+          expect($scope.list[0].id).toBe('1200');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Same as above.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 11]
```diff
@@ -143,8 +145,136 @@ define([
...
+        beforeEach(function () {
+          $scope.sortBy('assignee');
+        });
+
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+
+        it('lists documents by assignees', function () {
+          expect($scope.list[0].id).toBe('1200');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Same as above.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 12]
```diff
@@ -143,8 +145,136 @@ define([
...
+        beforeEach(function () {
+          $scope.sortBy('case_id');
+        });
+
+        afterEach(function () {
+          $rootScope.$apply();
+        });
+
+        it('lists documents by assignment type', function () {
+          expect($scope.list[0].id).toBe('1205');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Same as above.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js (hunk 1 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js]: removed!

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js (hunk 2 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/document-list.js]: right! fixed

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js (hunk 3 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: yes we can. Will fix it.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js (hunk 4 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js (hunk 5 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: done!

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js (hunk 6 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Re-written tests.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js (hunk 3 above)
Comment (deb1990) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: Can we use `contactIds = contactIds.concat(document.assignee_contact_id)`? which is more readable in my opinion.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js [hunk 13]
```diff
@@ -143,8 +145,140 @@ define([
...
+        beforeEach(function () {
+          sortedDocumentList = _.sortBy($scope.list, function (doc) {
+            return $rootScope.cache.documentStatus.obj[doc.status_id];
+          });
+          $scope.sortBy('status_id');
+        });
+
+        it('lists documents by document status', function () {
+          expect($scope.list).toEqual(sortedDocumentList);
+          expect($scope.list[4].id).toBe('1200');
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: Do we still need this?

Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/document-list.spec.js]: ops! will remove it.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js (hunk 3 above)
Comment (bmnepali) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/document.js]: sounds more appropriate, will do.

**Comment by bmnepali** [comment]
Fixed merge conflicts by rebase. Now merging in epic branch "PCHR-2112-onboarding-and-exiting-workflow"

**Pull request merged**